CONTRIBUTING.md

```diff
@@ -1,16 +1,90 @@
-Bug reports and feature requests for **the opam tool** should be reported on:
+Contributions are most welcome in any form.
 
+If you have bug reports and feature requests for the opam tool, please report them at:
 * http://github.com/ocaml/opam/issues
   (please include the output of `opam config report` whenever possible)
 
-**Packaging issues** or requests for a new package should be reported on:
-
+For packaging issues (e.g. packages failed to build) please report them at
 * http://github.com/ocaml/opam-repository/issues
 
-**General queries** can be addressed at:
+## Working on the opam codebase
+
+### Dependencies
+
+To get the dependencies required to build and test opam there are several ways depending on what you have currently.
+
+If you do not have OCaml already installed, you can simply call:
+```
+make cold
+```
+
+If you have OCaml but do not have opam already installed, you can simply call:
+```
+./configure --with-vendored-deps
+```
+
+If you have opam installed already, you can call:
+```
+opam install --deps-only -t .
+```
+
+### Building
+
+As long as you have the dependencies, compiling opam is as simple as:
+```
+./configure
+make
+```
+
+### Running the tests
+
+Once built, you can run the testsuite using:
+```
+make test
+```
+
+If you want to run one test in particular:
+```
+make reftest-<name of the test>
+```
+or
+```
+make reftest-<name of the test> DUNE_ARGS="--auto-promote"
+```
+if you want to regenerate a test based on the new output of opam.
```

Comment:
```suggestion
if you want to regenerate a test based on the new output of opam.
You can still cal afterwards `dune promote` to promote the new output.
```

Reply:
I'm not sure to understand that addition. The command above uses `--auto-promote` so `dune promote` shouldn't be necessary

Reply:
My proposal is not clear then :)
```suggestion
You can also run the test with `make reftest-<name of the test>, see the diff,
and call afterwards `dune promote` to promote the new output.
```